proposals/testing/NNNN-polling-confirmations.md

```diff
@@ -0,0 +1,512 @@
+closure returns `false` or a `nil` value. At this point, the confirmation will
+be marked as failing: an error will be thrown, and an issue will be recorded.
+
+Under both `PollingStopCondition` cases, when the early stop condition isn't
+reached, polling will continue up until approximately the `within`/`duration`
+value has elapsed. When `PollingStopCondition.firstPass` is specified, reaching
+the duration stop point will mark the confirmation as failing.
+When `PollingStopCondition.stopsPassing` is specified, reaching the duration
+stop point will mark the confirmation as passing.
+
+Tests will now be able to poll code updating in the background using either of
+the stop conditions:
+
+```swift
+let subject = Aquarium()
+Task {
+    await subject.raiseDolphins()
+}
+await confirmation(until: .firstPass) {
+    subject.dolphins.count == 1
+}
+```
+
+## Detailed design
+
+### New confirmation functions
+
+We will introduce 2 new members of the confirmation family of functions to the
+testing library:
+
+```swift
+/// Poll expression within the duration based on the given stop condition
+///
+/// - Parameters:
+///   - comment: An optional comment to apply to any issues generated by this
+///     function.
+///   - stopCondition: When to stop polling.
+///   - duration: The expected length of time to continue polling for.
+///     This value may not correspond to the wall-clock time that polling lasts
+///     for, especially on highly-loaded systems with a lot of tests running.
+///     If nil, this uses whatever value is specified under the last
+///     ``PollingUntilFirstPassConfigurationTrait`` or
+///     ``PollingUntilStopsPassingConfigurationTrait`` added to the test or
+///     suite.
+///     If no such trait has been added, then polling will be attempted for
+///     about 1 second before recording an issue.
+///     `duration` must be greater than 0.
+///   - interval: The minimum amount of time to wait between polling attempts.
+///     If nil, this uses whatever value is specified under the last
+///     ``PollingUntilFirstPassConfigurationTrait`` or
+///     ``PollingUntilStopsPassingConfigurationTrait`` added to the test or
+///     suite.
+///     If no such trait has been added, then polling will wait at least
+///     1 millisecond between polling attempts.
+///     `interval` must be greater than 0.
+///   - isolation: The actor to which `body` is isolated, if any.
+///   - sourceLocation: The source location to whych any recorded issues should
+///     be attributed.
+///   - body: The function to invoke.
+///
+/// - Throws: A `PollingFailedError` if the `body` does not return true within
+///   the polling duration.
+///
+/// Use polling confirmations to check that an event while a test is running in
+/// complex scenarios where other forms of confirmation are insufficient. For
+/// example, waiting on some state to change that cannot be easily confirmed
+/// through other forms of `confirmation`.
+@available(_clockAPI, *)
```

Comment:
`_clockAPI` is an availability definition specific to the Swift Testing codebase, and at build time it expands to actual OS version numbers. For the purpose of an API proposal I recommend showing the actual, expanded OS versions so people understand what they'll be in practice.

(This comment applies everywhere in the document which references `_clockAPI`.)